[fx] use a linked list for nodes

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45193 [fx] allow custom behavior for args, kwargs, and bool
* #46121 [fx] make sure args/kwargs are immutable
* **#45708 [fx] use a linked list for nodes**

This makes it possible to define reasonable semantics for what happens
when a node in the list is deleted. In particular the iteration over nodes
will continue at the node that was after the deleted node _when it was deleted_.
If the new node is also deleted, we skip it and, continue to the node after it.
Eventually we either reach a node still in the list or we reach the end of the list.

Differential Revision: [D24089516](https://our.internmc.facebook.com/intern/diff/D24089516)

zdevito committed:
[fx] use a linked list for nodes

This makes it possible to define reasonable semantics for what happens
when a node in the list is deleted. In particular the iteration over nodes
will continue at the node that was after the deleted node _when it was deleted_.
If the new node is also deleted, we skip it and, continue to the node after it.
Eventually we either reach a node still in the list or we reach the end of the list.

[ghstack-poisoned]

zdevito committed:
Update on "[fx] use a linked list for nodes"


This makes it possible to define reasonable semantics for what happens
when a node in the list is deleted. In particular the iteration over nodes
will continue at the node that was after the deleted node _when it was deleted_.
If the new node is also deleted, we skip it and, continue to the node after it.
Eventually we either reach a node still in the list or we reach the end of the list.

[ghstack-poisoned]

jamesr66a reviewed (commented):
Dang, this is really clean in Python compared to C++

zdevito commented on torch/fx/graph.py:
```diff
@@ -132,11 +136,23 @@ def erase_node(self, to_erase : Node):
+
+    def insert_after(self, n: Node):
+        assert n.graph == self, "Node to insert after is not in graph."
+        return _InsertPoint(self, n.append)
```
I change in to `inserting_before` so it read `with graph.inserting_before(n):` which should make it clear and not make it seem like `insert_before` is used to insert a node.

zdevito commented on torch/fx/node.py:
```diff
@@ -33,6 +33,32 @@ def __init__(self, graph: 'Graph', name: str, op: str, target: Target,
+    def append(self, x: 'Node'):
+        self._next.prepend(x)
+
+    def _remove_from_list(self):
```
the user never should see nodes that exist but are not in the list somewhere. The public function `erase_node` is the public API. This one doesn't set the _erased attribute because it also used when moving a node.

zdevito committed:
Update on "[fx] use a linked list for nodes"


This makes it possible to define reasonable semantics for what happens
when a node in the list is deleted. In particular the iteration over nodes
will continue at the node that was after the deleted node _when it was deleted_.
If the new node is also deleted, we skip it and, continue to the node after it.
Eventually we either reach a node still in the list or we reach the end of the list.

[ghstack-poisoned]

zdevito committed:
Update on "[fx] use a linked list for nodes"


This makes it possible to define reasonable semantics for what happens
when a node in the list is deleted. In particular the iteration over nodes
will continue at the node that was after the deleted node _when it was deleted_.
If the new node is also deleted, we skip it and, continue to the node after it.
Eventually we either reach a node still in the list or we reach the end of the list.

Differential Revision: [D24089516](https://our.internmc.facebook.com/intern/diff/D24089516)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] use a linked list for nodes"


This makes it possible to define reasonable semantics for what happens
when a node in the list is deleted. In particular the iteration over nodes
will continue at the node that was after the deleted node _when it was deleted_.
If the new node is also deleted, we skip it and, continue to the node after it.
Eventually we either reach a node still in the list or we reach the end of the list.

Differential Revision: [D24089516](https://our.internmc.facebook.com/intern/diff/D24089516)

[ghstack-poisoned]

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -153,11 +158,59 @@ def erase_node(self, to_erase : Node):
+    # these can be used to permanently set the insert point: g.insert_before(x)
+    # or when used in a with statement, to temporarily change the insert point:
+    # with g.insert_before(n):
+    #     <do some insertions>
```
nit: comment still says `insert_before`

Also, i think the docstring is sufficient here

jamesr66a commented on torch/fx/node.py:
```diff
@@ -51,6 +51,46 @@ def __init__(self, graph: 'Graph', name: str, op: str, target: Target,
+           Before: p -> self
+                   bx -> x -> ax
+           After:  p -> x -> self
+                   bx -> ax
```
It took me a second to figure out what the `bx` `ax` business meant. Maybe drop a line of prose at 67 saying "If `x` is already part of a node list, it is removed from that node list." or something?

jamesr66a reviewed (approved):
![image](https://user-images.githubusercontent.com/4685384/95524206-94ff6d00-0985-11eb-811f-2227584f03ad.png)

zdevito committed:
Update on "[fx] use a linked list for nodes"


This makes it possible to define reasonable semantics for what happens
when a node in the list is deleted. In particular the iteration over nodes
will continue at the node that was after the deleted node _when it was deleted_.
If the new node is also deleted, we skip it and, continue to the node after it.
Eventually we either reach a node still in the list or we reach the end of the list.

Differential Revision: [D24089516](https://our.internmc.facebook.com/intern/diff/D24089516)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] use a linked list for nodes"


This makes it possible to define reasonable semantics for what happens
when a node in the list is deleted. In particular the iteration over nodes
will continue at the node that was after the deleted node _when it was deleted_.
If the new node is also deleted, we skip it and, continue to the node after it.
Eventually we either reach a node still in the list or we reach the end of the list.

Differential Revision: [D24089516](https://our.internmc.facebook.com/intern/diff/D24089516)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] use a linked list for nodes"


This makes it possible to define reasonable semantics for what happens
when a node in the list is deleted. In particular the iteration over nodes
will continue at the node that was after the deleted node _when it was deleted_.
If the new node is also deleted, we skip it and, continue to the node after it.
Eventually we either reach a node still in the list or we reach the end of the list.

Differential Revision: [D24089516](https://our.internmc.facebook.com/intern/diff/D24089516)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] use a linked list for nodes"


This makes it possible to define reasonable semantics for what happens
when a node in the list is deleted. In particular the iteration over nodes
will continue at the node that was after the deleted node _when it was deleted_.
If the new node is also deleted, we skip it and, continue to the node after it.
Eventually we either reach a node still in the list or we reach the end of the list.

Differential Revision: [D24089516](https://our.internmc.facebook.com/intern/diff/D24089516)

[ghstack-poisoned]
